### PR TITLE
Implement StateDiff commitment calculation

### LINKED ISSUE #1090: Implement `state_diff_commitment` calculation
as described here https://community.starknet.io/t/introducing-p2p-authentication-and-mismatch-resolution-in-v0-12-2/97993

### PR DESCRIPTION
**Pull request by omerfirmak**
Fixes #1090

Ugly, but works :shrug:

### CODE REVIEW

File: core/state_update.go [hunk 1]
```diff
@@ -12,18 +14,18 @@ type StateUpdate struct {
...
+	ReplacedClasses   []AddressClassHashPair
 }
 
 type StorageDiff struct {
 	Key   *felt.Felt
 	Value *felt.Felt
 }
 
-type DeployedContract struct {
+type AddressClassHashPair struct {
```
Comment (IronGauntlets) [inline core/state_update.go:31]: A comment would be useful to describe what this type could represent.

File: core/state_update.go [hunk 2]
```diff
@@ -34,3 +37,108 @@ type DeclaredV1Class struct {
 	ClassHash         *felt.Felt
 	CompiledClassHash *felt.Felt
 }
+
+func (d *StateDiff) Commitment() *felt.Felt {
```
Comment (IronGauntlets) [inline core/state_update.go:41]: Where would this be used?

Comment (omerfirmak) [reply core/state_update.go:41]: We will use it to verify the signature from sequencer https://community.starknet.io/t/introducing-p2p-authentication-and-mismatch-resolution-in-v0-12-2/97993#signature-on-state-diff-commitment-and-block-hash-3

**Review by IronGauntlets — APPROVED** [review]
Please resolve the comments. 

lgtm.

File: core/state_update.go [hunk 3]
```diff
@@ -34,3 +37,108 @@ type DeclaredV1Class struct {
...
+	hashOfDeployedContracts.Update(tmpFelt.SetUint64(uint64(len(deployedReplacedContracts))))
+	sort.Slice(deployedReplacedContracts, func(i, j int) bool {
+		cmpRes := deployedReplacedContracts[i].Address.Cmp(deployedReplacedContracts[j].Address)
+		if cmpRes == -1 {
+			return true
+		} else if cmpRes == 1 {
+			return false
+		}
+		return deployedReplacedContracts[i].ClassHash.Cmp(deployedReplacedContracts[j].ClassHash) == -1
+	})
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply core/state_update.go:61]: There is nothing preventing that AFAICS.

**Comment by omerfirmak** [comment]
macos checks are failing but it cannot be due to this PR, force merging.

**Pull request merged**